### PR TITLE
Shared values

### PR DESCRIPTION
**Pull request by camaelon**

### CODE REVIEW

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java [hunk 1]
```diff
@@ -66,8 +66,59 @@
...
+     * @return
+     */
+    public int getSharedValue() {
+        return mSharedValue;
+    }
+
+    /**
+     * sets the SharedValue it will be listening for.
+     */
+    public void setSharedValue(int mSharedValue) {
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java:103]: rename mSharedValue parameter

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java [hunk 2]
```diff
@@ -100,6 +151,18 @@
...
+
+    public int getSharedValueCurrent() {
+        return mSharedValueCurrent;
+    }
+
+    public void setSharedValueCurrent(int sharedValueCurrent) {
+        this.mSharedValueCurrent = sharedValueCurrent;
+    }
+
+    private int mSharedValueCurrent = UNSET;
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java:165]: move with the other fields?

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java [hunk 3]
```diff
@@ -66,8 +66,59 @@
...
+    public static final int ONSTATE_ACTION_UP = 2;
+    public static final int ONSTATE_SHARED_VALUE_SET = 3;
+    public static final int ONSTATE_SHARED_VALUE_UNSET = 4;
+
+    /**
+     * Gets the type of transition to listen to.
+     *
+     * @return ONSTATE_TRANSITION_*
+     */
+    public int getStateTransition() {
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java:79]: should you move those functions after the fields?

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java [hunk 4]
```diff
@@ -306,6 +373,8 @@ void applyTransition(ViewTransitionController controller,
                          int fromId,
                          ConstraintSet current,
                          View... views) {
+        Log.v(TAG,Debug.getLoc()+"   "+Debug.getName(views[0]));
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java:376]: remove

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ReactiveGuide.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (C) 2015 The Android Open Source Project
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ReactiveGuide.java:2]: (c) 2020

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ReactiveGuide.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+import android.content.Context;
+import android.content.res.TypedArray;
+import android.graphics.Canvas;
+import android.util.AttributeSet;
+import android.view.View;
+
+import androidx.constraintlayout.motion.widget.MotionLayout;
+
+/**
+ * Utility class representing a Guideline helper object for {@link ConstraintLayout}.
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ReactiveGuide.java:29]: delete the javadoc for now

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/SharedValues.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2015 The Android Open Source Project
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/SharedValues.java:2]: (c) 2020

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java [hunk 8]
```diff
@@ -100,6 +151,18 @@
     private int mClearsTag = UNSET;
     private int mIfTagSet = UNSET;
     private int mIfTagNotSet = UNSET;
+    private int mSharedValue = UNSET;
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java:154]: maybe mSharedValueTarget?

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java [hunk 9]
```diff
@@ -100,6 +151,18 @@
...
+    private int mSharedValueID = UNSET;
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/motion/widget/ViewTransition.java:155]: before those variables maybe add something like:
// shared value management. mSharedValueId is the key we are watching,
// mSharedValueCurrent the current value for that key, and mSharedValueTarget
// is the target we are waiting for to trigger.

**Review by jafu888 — APPROVED** [review]

**Pull request merged**